### PR TITLE
[Inductor] Fix the decompositions of torch isin

### PR DESCRIPTION
**Pull request by leslie-fang-intel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #147519

**Summary**
Fixed two decomposition issues in `torch.isin`:

- Issue 1: As reported in [#147329](https://github.com/pytorch/pytorch/issues/147329), the current decomposition does not support cases where test_element is a scalar. This is now implemented by referring to the https://github.com/pytorch/pytorch/blob/ead970c8d035690c180641909b75da13fa16c76e/aten/src/ATen/native/TensorCompare.cpp#L1004-L1008

- Issue 2: Found while enabling a unit test with `elements = 1` and `test_elements = torch.tensor([1, 2, 3, 4])`, where Inductor produced different results compared to eager mode. This issue is fixed by referring to https://github.com/pytorch/pytorch/blob/ead970c8d035690c180641909b75da13fa16c76e/aten/src/ATen/native/cpu/TensorCompareKernel.cpp#L329-L338

**Test Plan**
```
python test/inductor/test_torchinductor.py -k test_isin_tensor_scalar
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

### CODE REVIEW

**Review by jgong5 — APPROVED** [review]

File: torch/_decomp/decompositions.py [hunk 1]
```diff
@@ -5091,7 +5091,10 @@ def isin(elements, test_elements, *, assume_unique=False, invert=False):
     if not isinstance(elements, torch.Tensor):
         elements = torch.tensor(elements, device=test_elements.device)
     if not isinstance(test_elements, torch.Tensor):
-        test_elements = torch.tensor(test_elements, device=elements.device)
+        if invert:
+            return torch.ne(elements, test_elements)
+        else:
+            return torch.eq(elements, test_elements)
```
Comment (peterbell10) [inline torch/_decomp/decompositions.py:5097]: Judging by the other change, this should be
```suggestion
        cmp = torch.eq(elements, test_elements)
        return ~cmp if invert else cmp
```

File: torch/_decomp/decompositions.py [hunk 2]
```diff
@@ -5123,14 +5126,10 @@ def bernoulli(
...
-
     x = elements.view(*elements.shape, *((1,) * test_elements.ndim))
-    if not invert:
-        cmp = x == test_elements
-    else:
-        cmp = x != test_elements
     dim = tuple(range(-1, -test_elements.ndim - 1, -1))
-    return cmp.any(dim=dim)
+    res = (x == test_elements).any(dim=dim)
+    return torch.where(res, not invert, invert)
```
Comment (peterbell10) [inline torch/_decomp/decompositions.py:5132]: NIT: could be simplified to
```suggestion
    return ~res if invert else res
```

Comment (leslie-fang-intel) [reply torch/_decomp/decompositions.py:5132]: Thanks for the suggestions! Changed.

File: torch/_decomp/decompositions.py (hunk 1 above)
Comment (leslie-fang-intel) [reply torch/_decomp/decompositions.py:5097]: Thanks for your suggestion. Since `test_elements` is a scalar tensor in this case, the issue fixed by the other change shouldn't occur here as using `ops.ne`. Additionally, when `invert` is True, the previous implementation uses only one `ops.ne`, whereas the new implementation introduces two operations: `ops.eq` and `ops.logical_not`. Therefore, I feel the previous implementation is more efficient. Looking forward to your further comments.

**Review by fffrog — APPROVED** [review]

File: torch/_decomp/decompositions.py (hunk 1 above)
Comment (peterbell10) [reply torch/_decomp/decompositions.py:5097]: I think this gives a different result for floating point nans, `(x == y) != !(x != y)` when either `x` or `y` is NaN.

Comment (leslie-fang-intel) [reply torch/_decomp/decompositions.py:5097]: > (x == y) != !(x != y) when either x or y is NaN.

Not sure if my understanding is right. But It seems to return `False` in C++, which appears to treat `(x == y)` as equivalent to `!(x != y)` when either `x` or `y` is `NaN`.

```
double x = std::nan("");
double y = 1.0;
std::cout<< ((x == y)) <<std::endl;
std::cout<< (!(x != y)) <<std::endl;
std::cout<< ((x == y) != !(x != y)) <<std::endl;
std::cout<< ((x == y) == !(x != y)) <<std::endl;
---------------

0
0
0
1
```

Comment (peterbell10) [reply torch/_decomp/decompositions.py:5097]: Ah okay, the expected condition doesn't hold for `<` and `>=`  and I extrapolated too far. So why did you make that change in `isin_default` then? What am I missing.

Comment (leslie-fang-intel) [reply torch/_decomp/decompositions.py:5097]: Ah, the previous implementation has a different result as eager when the `elements = 1`, `test_elements = torch.tensor([1, 2, 3, 4])` and `invert`

- `(x != test_elements).any(dim=dim)` returns `True`
- `~((x == test_elements).any(dim=dim))` returns `False`

Comment (peterbell10) [reply torch/_decomp/decompositions.py:5097]: Ah of course, because `~any(x) <=> all(~x)` and not `any(~x)`.

**Review by peterbell10 — APPROVED** [review]
